test/integration/agent/networkpolicy_metrics_test.go

```diff
@@ -0,0 +1,181 @@
+	return controller, clientset
+}
+
+func newNetworkPolicy(uid string, from, to, appliedTo []string, services []v1beta1.Service) *v1beta1.NetworkPolicy {
+	networkPolicyRule1 := v1beta1.NetworkPolicyRule{
+		Direction: v1beta1.DirectionIn,
+		From:      v1beta1.NetworkPolicyPeer{AddressGroups: from},
+		To:        v1beta1.NetworkPolicyPeer{AddressGroups: to},
+		Services:  services,
+	}
+	return &v1beta1.NetworkPolicy{
+		ObjectMeta:      v1.ObjectMeta{UID: types.UID(uid), Name: uid, Namespace: testNamespace},
+		Rules:           []v1beta1.NetworkPolicyRule{networkPolicyRule1},
+		AppliedToGroups: appliedTo,
+	}
+}
+
+func getNetworkPolicyWithMultipleRules(uid string, from, to, appliedTo []string, services []v1beta1.Service) *v1beta1.NetworkPolicy {
+	networkPolicyRule1 := v1beta1.NetworkPolicyRule{
+		Direction: v1beta1.DirectionIn,
+		From:      v1beta1.NetworkPolicyPeer{AddressGroups: from},
+		To:        v1beta1.NetworkPolicyPeer{AddressGroups: to},
+		Services:  services,
+	}
+	networkPolicyRule2 := v1beta1.NetworkPolicyRule{
+		Direction: v1beta1.DirectionOut,
+		From:      v1beta1.NetworkPolicyPeer{AddressGroups: from},
+		To:        v1beta1.NetworkPolicyPeer{AddressGroups: to},
+		Services:  services,
+	}
+	networkPolicyRule3 := v1beta1.NetworkPolicyRule{
+		Direction: v1beta1.DirectionOut,
+		From:      v1beta1.NetworkPolicyPeer{},
+		To:        v1beta1.NetworkPolicyPeer{AddressGroups: to},
+		Services:  services,
+	}
+	return &v1beta1.NetworkPolicy{
+		ObjectMeta:      v1.ObjectMeta{UID: types.UID(uid), Name: uid, Namespace: testNamespace},
+		Rules:           []v1beta1.NetworkPolicyRule{networkPolicyRule1, networkPolicyRule2, networkPolicyRule3},
+		AppliedToGroups: appliedTo,
+	}
+}
+
+func TestAddMultipleNewtorkPolicy(t *testing.T) {
+	// Initialize NetworkPolicy metrics (prometheus)
+	metrics.InitializeNetworkPolicyMetrics()
+
+	controller, clientset := newTestController()
+	addressGroupWatcher := watch.NewFake()
+	appliedToGroupWatcher := watch.NewFake()
+	networkPolicyWatcher := watch.NewFake()
+	clientset.AddWatchReactor("addressgroups", k8stesting.DefaultWatchReactor(addressGroupWatcher, nil))
+	clientset.AddWatchReactor("appliedtogroups", k8stesting.DefaultWatchReactor(appliedToGroupWatcher, nil))
+	clientset.AddWatchReactor("networkpolicies", k8stesting.DefaultWatchReactor(networkPolicyWatcher, nil))
+
+	protocolTCP := v1beta1.ProtocolTCP
+	port := intstr.FromInt(80)
+	services := []v1beta1.Service{{Protocol: &protocolTCP, Port: &port}}
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+	go controller.Run(stopCh)
+
+	// Test adding policy1 with a single rule
+	policy1 := newNetworkPolicy("policy1", []string{"addressGroup1"}, []string{}, []string{"appliedToGroup1"}, services)
+	networkPolicyWatcher.Add(policy1)
+	networkPolicyWatcher.Action(watch.Bookmark, nil)
+	// Wait 100ms for metrics update
+	time.Sleep(time.Millisecond * 100)
```

Comment:
Just wondering if this expected delay is inline with the time elapsed in following lines?
func (c *Controller) syncRule(key string) error {
	startTime := time.Now()
	defer func() {
		klog.V(4).Infof("Finished syncing rule %q. (%v)", key, time.Since(startTime))
	}()

Reply:
I'll check the expected delay.

Reply:
Just for your information, here is the delay in my test
```
I0730 15:57:49.610568     167 networkpolicy_controller.go:439] Received 1 init events for NetworkPolicy
I0730 15:57:49.610575     167 networkpolicy_controller.go:139] NetworkPolicy ns1/policy1 applied to Pods on this Node
E0730 15:57:49.610848     167 networkpolicy_controller.go:335] Finished syncing rule "3812408e464bed71". (2.243µs)
W0730 15:57:49.656048     167 interface_configuration_linux.go:78] Failed to send gratuitous ARP #2: no such device or address
I0730 15:57:49.716261     167 networkpolicy_controller.go:111] NetworkPolicy ns1/policy2 applied to Pods on this Node
E0730 15:57:49.716339     167 networkpolicy_controller.go:335] Finished syncing rule "e3d3e37daf0d7aeb". (35.009µs)
E0730 15:57:49.716354     167 networkpolicy_controller.go:335] Finished syncing rule "782bb947897a7c96". (2.346µs)
E0730 15:57:49.716364     167 networkpolicy_controller.go:335] Finished syncing rule "793ef4529d0e2427". (1.252µs)
I0730 15:57:49.821300     167 networkpolicy_controller.go:128] NetworkPolicy ns1/policy1 no longer applied to Pods on this Node
I0730 15:57:49.821355     167 reconciler.go:497] Forgetting rule 3812408e464bed71
E0730 15:57:49.821364     167 networkpolicy_controller.go:335] Finished syncing rule "3812408e464bed71". (11.803µs)
I0730 15:57:49.929005     167 networkpolicy_controller.go:128] NetworkPolicy ns1/policy2 no longer applied to Pods on this Node
I0730 15:57:49.929135     167 reconciler.go:497] Forgetting rule e3d3e37daf0d7aeb
E0730 15:57:49.929147     167 networkpolicy_controller.go:335] Finished syncing rule "e3d3e37daf0d7aeb". (14.546µs)
I0730 15:57:49.929160     167 reconciler.go:497] Forgetting rule 782bb947897a7c96
E0730 15:57:49.929164     167 networkpolicy_controller.go:335] Finished syncing rule "782bb947897a7c96". (4.993µs)
I0730 15:57:49.929171     167 reconciler.go:497] Forgetting rule 793ef4529d0e2427
E0730 15:57:49.929176     167 networkpolicy_controller.go:335] Finished syncing rule "793ef4529d0e2427". (4.696µs)
```

100 ms seems long enough to realize 1 networkpolicy.

Reply:
Thanks for the output. Sounds good. Looking at those numbers.. may be 1ms or 5ms good enough? Just incase if more network policy tests are added in future. Delay incurred to run integration tests may increase unnecessarily.

Reply:
please add a polling mechanism if possible instead to make the test more robust

https://godoc.org/k8s.io/apimachinery/pkg/util/wait

there are several examples in the e2e test directory

Reply:
Thank you for your suggestion. That makes sense.
I use the below interval and timeout for polling.
```
const networkPolicyTimeout = time.Millisecond * 100
const networkPolicyInterval = time.Millisecond * 5
```